Feature/cargo push

jsvisa opened the pull request:
@ddysher @ijumps PTAL

jsvisa committed:
feat(make): copy binary to docker

        DOCKER_REGISTRY=cargo.caicloudprivatetest.com make docker

jsvisa committed:
feat(docker): dockerfile and start script

deyuanix commented:
@ijump

ijumps commented on Dockerfile:
```diff
@@ -0,0 +1,17 @@
+FROM ubuntu:16.04
+MAINTAINER Delweng Zheng <delweng@gmail.com>
+
+RUN apt-get update && apt-get install -y cron
```
multi `RUN` to one `RUN` to reduce docker image layer, e.g.

```
##update system
RUN apt-get update && apt-get upgrade -y \

##runtime install
	&& apt-get install -y nginx \

##clean
	&& apt-get autoremove -y \
	&& rm -rf /var/lib/apt/lists/*
```

jsvisa replied:
@ijumps fixed

ijumps commented on heketi-start.sh:
```diff
@@ -0,0 +1,59 @@
+#!/bin/sh
```
If you use a bash script as an entrypoint, a good practice is to write script like this:

```
#!/bin/bash
set -e

#some setup here

exec "$@"
```

A shell script will not handle linux signal well, and this shell will become pid 1. call `exec` will let the real entrypoint as pid 1

jsvisa replied:
@ijumps  thanks for advise

ijumps commented on heketi-start.sh:
```diff
@@ -0,0 +1,59 @@
+            datadb=$2; shift 2;;
+        --)
+            shift; break;;
+        -h|--help)
```
`-h` not work

jsvisa replied:
@ijumps fixed

ijumps commented on heketi-start.sh:
```diff
@@ -0,0 +1,59 @@
+mkdir -p $backup
+if [ ! -f $datadb ]; then
+    recent=$(ls $backup | sort -r | head -1)
+    if [ $recent -ne "" ]; then
```
`$recent -ne ""` to `"$recent" != ""`

jsvisa replied:
@ijumps thanks for advise

ijumps commented on Makefile:
```diff
@@ -90,6 +90,15 @@ clean:
+	docker build -t $(DOCKER_REGISTRY)/caicloud/$(APP_NAME):$(VERSION:v%=%) .
+else
+	@echo "Push image $(APP_NAME):$(VERSION) to default docker registry(default is hub.docker.com)"
+	docker build -t caicloud/$(APP_NAME):$(VERSION:v%=%) .
```
we usually keep image prefix `v`, e.g. `v1.2.3`

jsvisa replied:
fixed @ijumps

ijumps commented on Makefile:
```diff
@@ -90,6 +90,15 @@ clean:
 
+docker:
+ifdef DOCKER_REGISTRY
+	@echo "Push image $(APP_NAME):$(VERSION) to $(DOCKER_REGISTRY)/caicloud/$(APP_NAME):$(VERSION:v%=%)"
```
this only do docker build but not docker push

jsvisa replied:
@ijumps added

jsvisa committed:
fix(docker): move multi RUN into one

jsvisa committed:
fix(docker): bash syntax in init script

1. fix: `-h` not working;
2. fix: $recent -ne "" -> "$recent" != ""
3. improve: exec command make the entrypoint to be pid 1

jsvisa committed:
feat(make): push docker image after build

deyuanix commented:
@ijumps can we merge can proceed?

ijumps commented on heketi-start.sh:
```diff
@@ -0,0 +1,59 @@
+    recent=$(ls $backup | sort -r | head -1)
+    if [ "$recent" != "" ]; then
+        tar zxf $backup/$recent -C $datadb_dir
+        ret=$?
```
any error will cause script exit when `set -e`, for this script, `ret=$?` will never reach if `tar zxf $backup/$recent -C $datadb_dir` failed. So `set -e` may not need here.

jsvisa replied:
@ijumps  fixed  https://github.com/caicloud/heketi/commit/ff0551823f03bced72da13892d48b3e5ea9f97dd

ijumps commented on heketi-start.sh:
```diff
@@ -0,0 +1,59 @@
+echo "30 2 * * * root (cd $datadb_dir; export backup_file=heketi-db-\$(date +\\%F-\\%H-\\%M).tar.gz && tar zcf \$backup_file heketi.db && mv \$backup_file $backup) >/dev/null 2>&1\n#" > /etc/crontab
+
+pidof cron
+if [ $? != 0 ]; then
```
the same

jsvisa committed:
fix(docker): 'set -e' means exit immediatelly if error occured

ijumps commented:
LGTM
